doc/compiler/tril/TrilLanguageReference.md

```diff
@@ -0,0 +1,194 @@
+# Tril Language Reference
+
+In may ways, the Tril language mirrors the Testarossa Intermediate Language,
+_Trees_. 
+
+This language reference intends to document mostly the places where special
+behaviour has been created in Tril to make writing test cases easier. 
+
+# Basics
+
+Tril is built out of two main components: 
+
+1. A very permissive S-expression parser.
+2. An IlGenerator that walks the AST produced by said parser, and generates
+   trees. 
+
+# Syntax
+
+Tril represents Trees using S-expressions, with key-value properties to augment
+a particular node with more information. 
+
+* `(x (y) (z))` represents some `x` with `y` and `z` as children.
+* `(x foo=bar baz=braz (y) (z)` is the same tree as above, but where the node
+  for `x` is augmented with two properties with values. 
+
+  Anything may be provided as a property for a node, however, unknown properties will be silently ignored.
+
+
+## Special Forms: 
+
+The key to Tril is the unlocking of extra behaviour using 'special forms', or
+atoms that represent certain behaviours in Testarossa. 
+
+### Methods
+
+A whole method in Tril is wrapped in a `method` expression. 
+
+#### Properties 
+
+* `return` _Mandatory_ is the return type of the method. 
+* `args` _Optional_ is a square-bracketed list of arguments representing the
+  types of the parameters, and the order in which they appear. 
+
+  eg. `args=[Int32,Double,Float]`
+    
+
+
+### Block
+
+In the Testarossa IL, all trees exist within the context of a basic block. 
+
+#### Properties
+
+* `name` _Optional_ Blocks can be named in order to target them with branches. 
+
+### Stores and loads
+
+Currently the creation of new symbol references in Tril is not supported.
+However, store and load opcodes provide two options for where the store / load
+can come from, which can be specified with one of the following two proprties: 
+
+* `parm` can specify the parameter from which a load can come. eg. `(iload
+  parm=2)` loads the third parameter from the signature. 
+* `temp` Stores to temps create new temporary variables that can later be
+  referenced. eg. `(istore temp="x" ...)`
+
+### Branches 
+
+Branches evaluate a condition then either jump to a target, or falls through. 
```

Comment:
Not sure if documenting the behavior of opcodes is appropriate in this document. It really should be part of a document on Trees.

Although, I can see how something short like this can be helpful for readability and clarity.

Reply:
I think I'd like to keep this just because it increases the readability of the document.

Reply:
Does this need to consider table lookups and switches? what about goto - ie unconditional branch?

Reply:
Unconditional branches work the same way, so that phrase should be generalized to include these as well

Based on my weak understanding of table lookups and switches, I think they should also work, but I've never tested them myself.

Reply:
I think I'd keep switches as undocumented until we have a test case that's verifying it all works the way it ought to :)

Reply:
https://github.com/eclipse/omr/issues/2000